File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  bool isValidBST(TreeNode? root) {
    if(root == null) return true;
    return solve(root, double.infinity, double.negativeInfinity);
  }

  bool solve(TreeNode? root, double max, double min){
    if(root == null) return true;
    else if(root.val <= min || root.val >= max) return false;
    return solve(root.left, root.val.toDouble(), min) && solve(root.right, max, root.val.toDouble());
  }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return find(root, Long.MAX_VALUE, Long.MIN_VALUE);
    }

    public boolean find(TreeNode root, double max, double min){
        if(root == null) return true;
        else if(root.val <= min || root.val >= max) return false;
        return find(root.left, root.val, min) && find(root.right, max, root.val);
    }
}